<a href="https://colab.research.google.com/github/avellozo/DIO/blob/main/Carrefour/Gapminder.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Análise do dataset Gapminder

# Imports e inicializações gerais


In [78]:
import pandas as pd
pd.options.display.float_format = '{:.2f}'.format


#Dataset load

In [2]:
df = pd.read_csv("/content/drive/MyDrive/Colab Notebooks/DIO/Carrefour Data Engineer/Cusro_Python_Pandas_Digital_Innovation-master/datasets/Gapminder.csv", sep=";")

#Exploração do conteúdo do dataset

In [49]:
df.sample(10)

country continent  year  lifeExp        pop     gdpPercap
1635       Korea, Dem. Rep.      Asia  1992   69.978   20711375   3726.063507
707           Cote d'Ivoire    Africa  1987   54.655   10761098   2156.956069
3256                Vietnam      Asia  1972   50.254   44655014    699.501644
1557                  Japan      Asia  1966   71.140   99790308   8957.280440
2547           Saudi Arabia      Asia  1977   58.690    8128505  34167.762600
1198                Grenada       NaN  1992   66.469      91487   6098.922790
3215          United States  Americas  2005   77.890  295734134  41674.000000
1811             Luxembourg    Europe  1996   76.500     415566  49005.588530
2530  Sao Tome and Principe       NaN  1952   46.471      60011    879.583586
325                   Benin    Africa  1982   50.904    3641603   1277.897616

In [4]:
df.shape

(3312, 6)

Data set tem 3312 registros e 6 colunas. As colunas tem os seguintes tipos de dados:

In [5]:
df.dtypes

country       object
continent     object
year           int64
lifeExp      float64
pop            int64
gdpPercap    float64
dtype: object

#Análise preliminar

##Valores nulos e categóricos

Quais colunas tem valores nulos?

In [25]:
df.isna().any()

country      False
continent     True
year         False
lifeExp      False
pop          False
gdpPercap    False
dtype: bool

In [52]:
df.isna().sum()

country        0
continent    188
year           0
lifeExp        0
pop            0
gdpPercap      0
dtype: int64

In [44]:
df[df.continent.isna()].country.value_counts()

Hong Kong, China         12
Reunion                  12
Sao Tome and Principe    12
Haiti                    12
Belize                   11
Barbados                 10
Bahamas                  10
Papua New Guinea         10
French Polynesia          9
New Caledonia             9
Georgia                   9
Micronesia, Fed. Sts.     8
Grenada                   8
Aruba                     8
Netherlands Antilles      8
Maldives                  8
Vanuatu                   7
Samoa                     7
Tonga                     7
Armenia                   4
Uzbekistan                4
Guadeloupe                1
Martinique                1
French Guiana             1
Name: country, dtype: int64

Quais são os possíveis valores de continente e países?

In [28]:
df.continent.value_counts()

Europe      1302
Africa       613
Asia         557
Americas     343
FSU          122
Oceania       74
Name: continent, dtype: int64

In [19]:
df.country.value_counts()

Slovak Republic    58
Switzerland        58
Norway             58
Japan              58
Finland            58
                   ..
Kazakhstan          4
Uzbekistan          4
Martinique          1
French Guiana       1
Guadeloupe          1
Name: country, Length: 187, dtype: int64

##Arrumar os dados mais relevantes

In [69]:
df.loc[df.country == 'Canada','continent'] = 'Americas'
df.loc[df.country == 'Australia','continent'] = 'Oceania'
df.loc[df.country == 'Hong Kong, China','continent'] = 'Asia'
df.loc[df.country == 'Sao Tome and Principe','continent'] = 'Africa'
df.loc[df.country == 'Reunion','continent'] = 'Africa'
df.loc[df.country == 'Haiti','continent'] = 'Americas'
df.loc[df.country == 'Belize','continent'] = 'Americas'
df.loc[df.country == 'Barbados','continent'] = 'Americas'
df.loc[df.country == 'Bahamas','continent'] = 'Americas'
df.loc[df.country == 'Papua New Guinea','continent'] = 'Oceania'

# poderia acertar os outros países que são nulos tbm...

##Sumarização das colunas que são numéricas (describe)

In [77]:
df.describe()

year  lifeExp           pop  gdpPercap
count 3312.00  3312.00       3312.00    3312.00
mean  1980.30    65.25   31614890.82   11317.12
std     16.93    11.77  104119342.89   11369.14
min   1950.00    23.60      59412.00     241.17
25%   1967.00    58.34    2678572.00    2514.63
50%   1982.00    69.61    7557218.50    7838.51
75%   1996.00    73.66   19585221.75   17357.88
max   2007.00    82.67 1318683096.00  113523.13

Primeiras observações:
- Média da expectativa de vida: 65,24 anos e mediana 69,61. Mostra que existem muitos valores muito baixos
- População tem desvio padrão maior que a média, ou seja, valores estão muito dispersos. 
- PIB per capta também tem grande dispersão.
- Média e mediana do PIB per capta muito diferente. Devem haver muitos dados com valores muito altos. 

Perguntas diretas que saem desta primeira análise:
- Qual registro (país e ano) tem:
    - a menor expectativa de vida?
    - o maior e menor PIB per capta
    - a menor população
- Muitos registros com alta população?
- Existe alguma correlação entre países pequenos (pouca população) e expectativa de vida?
- Qual a correlação entre PIB per capta e expectativa de vida?

#Resultados das perguntas iniciais


#Análise da expectativa de vida

## Extremos da expectativa de vida

In [10]:
df[df.lifeExp == min(df.lifeExp)]

country continent  year  lifeExp      pop   gdpPercap
2519  Rwanda    Africa  1992   23.599  7290203  737.068595

In [11]:
df[df.lifeExp == max(df.lifeExp)]

country continent  year  lifeExp        pop    gdpPercap
1597   Japan      Asia  2006    82.67  127463611  31001.71797

Pergunta: será que Ruanda é o que tem as mais baixas expectativas de vida?
E as maiores expectativas de vida são do Japão?

In [70]:
df.sort_values(by='lifeExp').head(20)['country'].value_counts()

Afghanistan      3
Sierra Leone     3
Guinea-Bissau    2
Angola           2
Gambia           2
Rwanda           1
Yemen, Rep.      1
Guinea           1
Somalia          1
Mozambique       1
China            1
Burkina Faso     1
Cambodia         1
Name: country, dtype: int64

Apesar de Ruanda em 1952 ter tido a menor expectativa de vida, as 20 menores estão bem ditribuídas

In [71]:
df.sort_values(by='lifeExp').tail(20)['country'].value_counts()

Japan               8
Switzerland         4
Iceland             3
Australia           2
Hong Kong, China    2
Italy               1
Name: country, dtype: int64

Já o Japão aparece muitas vezes entre os 20 maiores

e se for contar por continente?


In [73]:
df.sort_values(by='lifeExp').head(20)['continent'].value_counts()

Africa    14
Asia       6
Name: continent, dtype: int64

In [74]:
df.sort_values(by='lifeExp').tail(20)['continent'].value_counts()

Asia       10
Europe      8
Oceania     2
Name: continent, dtype: int64

Asia e Europa estão equilibradas

## Quais países tiveram decréscimos na expectativa de vida com o passar dos anos?

In [106]:
df_country_year = df.sort_values(by=['country', 'year'])
df_country_year['diff'] = df_country_year.groupby('country')['lifeExp'].diff()
df_country_year[df_country_year['diff']<0]['country'].value_counts()

Iceland       23
Latvia        22
Bulgaria      22
Hungary       20
Portugal      17
              ..
Uzbekistan     1
Somalia        1
Eritrea        1
Ghana          1
Togo           1
Name: country, Length: 89, dtype: int64

Vários países tiveram pelo menos uma vez um decréscimo de expectativa de vida. Por exemplo Portugal e Iceland

In [111]:
df_country_year[df_country_year['country']=='Iceland']

country continent  year  lifeExp     pop  gdpPercap  diff
1342  Iceland    Europe  1950    71.01  142938    7750.29   nan
1343  Iceland    Europe  1951    71.05  145604    7468.81  0.04
1344  Iceland    Europe  1952    72.49  147962    7267.69  1.44
1345  Iceland    Europe  1953    72.31  151036    8211.29 -0.18
1346  Iceland    Europe  1954    73.36  154563    8800.72  1.05
1347  Iceland    Europe  1955    73.30  158044    9458.67 -0.06
1348  Iceland    Europe  1956    72.98  161358    9489.02 -0.32
1349  Iceland    Europe  1957    73.47  165110    9244.00  0.49
1350  Iceland    Europe  1958    73.43  168771    9789.87 -0.04
1351  Iceland    Europe  1959    72.66  172314    9768.93 -0.77
1352  Iceland    Europe  1960    74.09  175860    9890.07  1.43
1353  Iceland    Europe  1961    73.47  178905    9724.49 -0.62
1354  Iceland    Europe  1962    73.68  182053   10350.16  0.21
1355  Iceland    Europe  1963    72.98  185481   11221.63 -0.70
1356  Iceland    Europe  1964    73.57  188847   12133.87  0.59
1357  Iceland    Europe  1965    73.84  192288   12794.88  0.27
1358  Iceland    Europe  1966    73.23  195610   13701.27 -0.61
1359  Iceland    Europe  1967    73.73  198676   13319.90  0.50
1360  Iceland    Europe  1968    73.94  201245   12441.44  0.21
1361  Iceland    Europe  1969    73.71  202920   12659.11 -0.23
1362  Iceland    Europe  1970    73.81  204104   13532.84  0.10
1363  Iceland    Europe  1971    73.55  206092   15128.83 -0.26
1364  Iceland    Europe  1972    74.46  209275   15798.06  0.91
1365  Iceland    Europe  1973    74.31  212364   16670.02 -0.15
1366  Iceland    Europe  1974    74.37  215324   17375.59  0.06
1367  Iceland    Europe  1975    75.35  218031   17297.44  0.98
1368  Iceland    Europe  1976    76.71  220133   18179.25  1.36
1369  Iceland    Europe  1977    76.11  221823   19654.96 -0.60
1370  Iceland    Europe  1978    76.42  223587   20649.68  0.31
1371  Iceland    Europe  1979    76.63  225749   21453.78  0.21
1372  Iceland    Europe  1980    76.66  228161   22434.86  0.03
1373  Iceland    Europe  1981    76.35  230803   23121.41 -0.31
1374  Iceland    Europe  1982    76.99  233997   23269.61  0.64
1375  Iceland    Europe  1983    76.63  237041   22499.90 -0.36
1376  Iceland    Europe  1984    77.43  239498   23215.83  0.80
1377  Iceland    Europe  1985    77.56  241403   23821.71  0.13
1378  Iceland    Europe  1986    78.09  242999   25107.72  0.53
1379  Iceland    Europe  1987    77.23  244676   26923.21 -0.86
1380  Iceland    Europe  1988    77.08  249885   26435.22 -0.15
1381  Iceland    Europe  1989    78.15  252746   26317.46  1.07
1382  Iceland    Europe  1990    77.98  254719   26372.88 -0.17
1383  Iceland    Europe  1991    77.94  256831   26332.62 -0.04
1384  Iceland    Europe  1992    78.77  259012   25144.39  0.83
1385  Iceland    Europe  1993    78.98  261270   25030.86  0.21
1386  Iceland    Europe  1994    79.24  266075   25941.54  0.26
1387  Iceland    Europe  1995    78.05  267527   25840.49 -1.19
1388  Iceland    Europe  1996    78.84  269000   27027.25  0.79
1389  Iceland    Europe  1997    78.95  271192   28061.10  0.11
1390  Iceland    Europe  1998    79.67  274000   29256.70  0.72
1391  Iceland    Europe  1999    79.46  277280   30042.94 -0.21
1392  Iceland    Europe  2000    79.72  281043   31092.03  0.26
1393  Iceland    Europe  2001    80.64  284812   31830.54  0.92
1394  Iceland    Europe  2002    80.50  288030   31163.20 -0.14
1395  Iceland    Europe  2003    81.06  291064   31937.01  0.56
1396  Iceland    Europe  2004    80.95  293966   34279.18 -0.11
1397  Iceland    Europe  2005    81.43  296737   35630.00  0.48
1398  Iceland    Europe  2006    81.12  299388   36233.27 -0.31
1399  Iceland    Europe  2007    81.76  301931   36180.79  0.64

In [110]:
df_country_year[df_country_year['country']=='Portugal']

country continent  year  lifeExp       pop  gdpPercap  diff
2388  Portugal    Europe  1950    58.53   8442750    2961.89   nan
2389  Portugal    Europe  1951    58.72   8490250    3077.24  0.19
2390  Portugal    Europe  1952    59.82   8526050    3068.32  1.10
2391  Portugal    Europe  1953    61.12   8578950    3262.20  1.30
2392  Portugal    Europe  1954    62.26   8632100    3397.69  1.14
2393  Portugal    Europe  1955    61.43   8692600    3513.19 -0.83
2394  Portugal    Europe  1956    61.24   8756000    3639.99 -0.19
2395  Portugal    Europe  1957    61.51   8817650    3774.57  0.27
2396  Portugal    Europe  1958    63.81   8888550    3793.66  2.30
2397  Portugal    Europe  1959    62.99   8961550    3966.47 -0.82
2398  Portugal    Europe  1960    64.25   9036700    4196.15  1.26
2399  Portugal    Europe  1961    62.87   9031200    4428.04 -1.38
2400  Portugal    Europe  1962    64.39   9019800    4727.95  1.52
2401  Portugal    Europe  1963    65.02   9081600    4974.50  0.63
2402  Portugal    Europe  1964    65.24   9122500    5278.68  0.22
2403  Portugal    Europe  1965    66.19   9128850    5667.66  0.95
2404  Portugal    Europe  1966    65.70   9108800    5911.27 -0.49
2405  Portugal    Europe  1967    66.60   9103000    6361.52  0.90
2406  Portugal    Europe  1968    66.91   9115050    6918.30  0.31
2407  Portugal    Europe  1969    66.52   9097200    7079.03 -0.39
2408  Portugal    Europe  1970    67.17   9044200    7769.41  0.65
2409  Portugal    Europe  1971    66.94   8990450    8334.25 -0.23
2410  Portugal    Europe  1972    69.26   8970450    9022.25  2.32
2411  Portugal    Europe  1973    68.66   8975950   10026.71 -0.60
2412  Portugal    Europe  1974    69.21   9098300   10005.00  0.55
2413  Portugal    Europe  1975    68.93   9411090    9251.92 -0.28
2414  Portugal    Europe  1976    69.16   9621970    9673.53  0.23
2415  Portugal    Europe  1977    70.41   9662600   10172.49  1.25
2416  Portugal    Europe  1978    70.87   9698780   10419.96  0.46
2417  Portugal    Europe  1979    71.68   9724560   10978.31  0.81
2418  Portugal    Europe  1980    71.75   9777800   11419.72  0.07
2419  Portugal    Europe  1981    71.94   9850079   11519.39  0.19
2420  Portugal    Europe  1982    72.77   9859650   11753.84  0.83
2421  Portugal    Europe  1983    72.69   9872243   11718.43 -0.08
2422  Portugal    Europe  1984    72.98   9885387   11482.99  0.29
2423  Portugal    Europe  1985    73.26   9897192   11791.30  0.28
2424  Portugal    Europe  1986    73.66   9907411   12266.89  0.40
2425  Portugal    Europe  1987    74.06   9915289   13039.31  0.40
2426  Portugal    Europe  1988    74.07   9920611   14008.38  0.01
2427  Portugal    Europe  1989    74.63   9923147   14724.11  0.56
2428  Portugal    Europe  1990    74.25   9922689   15369.31 -0.38
2429  Portugal    Europe  1991    74.19   9919009   16046.61 -0.06
2430  Portugal    Europe  1992    74.86   9927680   16207.27  0.67
2431  Portugal    Europe  1993    74.77   9967832   15812.16 -0.09
2432  Portugal    Europe  1994    75.76  10027688   15869.43  0.99
2433  Portugal    Europe  1995    75.58  10065543   16486.84 -0.18
2434  Portugal    Europe  1996    75.53  10099918   17027.45 -0.05
2435  Portugal    Europe  1997    75.97  10156415   17641.03  0.44
2436  Portugal    Europe  1998    76.18  10224828   18368.52  0.21
2437  Portugal    Europe  1999    76.35  10283381   18959.84  0.17
2438  Portugal    Europe  2000    76.83  10335597   19609.73  0.48
2439  Portugal    Europe  2001    77.11  10386753   19904.87  0.28
2440  Portugal    Europe  2002    77.29  10433867   19970.91  0.18
2441  Portugal    Europe  2003    77.48  10479955   19727.84  0.19
2442  Portugal    Europe  2004    78.31  10524145   19942.44  0.83
2443  Portugal    Europe  2005    78.18  10566212   20006.00 -0.13
2444  Portugal    Europe  2006    78.92  10605870   20149.08  0.74
2445  Portugal    Europe  2007    78.10  10642836   20509.65 -0.82

##Alguns gráficos que demonstram a expectativa de vida com o tempo

#Análise da População

#Análise do PIB per capita

#Correlação entre população e expectativa de vida

#Correlação entre PIB per capta e expectativa de vida